<!DOCTYPE html>
<html lang="en">

<!-- Head tag -->
<head>

    <meta charset="utf-8">
    <meta http-equiv="X-UA-Compatible" content="IE=edge">
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />

    <!--Description-->
    
        <meta name="description" content="123456789101112131415161718192021222324252627282930313233343536373839404142434445464748495051525354555657585960616263646566676869707172737475767778798">
    

    <!--Author-->
    
        <meta name="author" content="Hebe">
    

    <!--Open Graph Title-->
    
        <meta property="og:title" content="Coding-Demonstration"/>
    

    <!--Open Graph Site Name-->
    <meta property="og:site_name" content="Hebe&#39;s blog"/>

    <!--Page Cover-->
    
        <meta property="og:image" content=""/>
    

    <!-- Title -->
    
    <title>Coding-Demonstration - Hebe&#39;s blog</title>

    <!-- Custom CSS -->
    <link rel="stylesheet" href="/sass/main.css">

    <!--[if lt IE 8]>
        <script src="/js/ie/html5shiv.js"></script>
    <![endif]-->

    <!--[if lt IE 8]>
        <link rel="stylesheet" href="/sass/ie8.css">
    <![endif]-->

    <!--[if lt IE 9]>
        <link rel="stylesheet" href="/sass/ie9.css">
    <![endif]-->

    <!-- Gallery -->
    <link href="//cdn.rawgit.com/noelboss/featherlight/1.3.5/release/featherlight.min.css" type="text/css" rel="stylesheet" />

    <!-- Google Analytics -->
    


</head>

<body>

    <div id="wrapper">

        <!-- Menu -->
        <!-- Header -->
<header id="header">
    <div class="inner">

        <!-- Logo -->
        <a href="/" class="logo">
            <span class="symbol"><img src="/images/logo.svg" alt="" /></span><span class="title">Hebe's blog</span>
        </a>

        <!-- Nav -->
        <nav>
            <ul>
                <li><a href="#menu">Menu</a></li>
            </ul>
        </nav>

    </div>
</header>

<!-- Menu -->
<nav id="menu">
    <h2>Menu</h2>
    <ul>
        
            <li>
                <a href="/">Home</a>
            </li>
        
    </ul>
</nav>


        <div id="main">
            <div class="inner">

                <!-- Main Content -->
                

    <h1>Coding-Demonstration</h1>



<!-- Gallery -->


<!-- Content -->
<figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br><span class="line">44</span><br><span class="line">45</span><br><span class="line">46</span><br><span class="line">47</span><br><span class="line">48</span><br><span class="line">49</span><br><span class="line">50</span><br><span class="line">51</span><br><span class="line">52</span><br><span class="line">53</span><br><span class="line">54</span><br><span class="line">55</span><br><span class="line">56</span><br><span class="line">57</span><br><span class="line">58</span><br><span class="line">59</span><br><span class="line">60</span><br><span class="line">61</span><br><span class="line">62</span><br><span class="line">63</span><br><span class="line">64</span><br><span class="line">65</span><br><span class="line">66</span><br><span class="line">67</span><br><span class="line">68</span><br><span class="line">69</span><br><span class="line">70</span><br><span class="line">71</span><br><span class="line">72</span><br><span class="line">73</span><br><span class="line">74</span><br><span class="line">75</span><br><span class="line">76</span><br><span class="line">77</span><br><span class="line">78</span><br><span class="line">79</span><br><span class="line">80</span><br><span class="line">81</span><br><span class="line">82</span><br><span class="line">83</span><br><span class="line">84</span><br><span class="line">85</span><br><span class="line">86</span><br><span class="line">87</span><br><span class="line">88</span><br><span class="line">89</span><br><span class="line">90</span><br><span class="line">91</span><br><span class="line">92</span><br><span class="line">93</span><br><span class="line">94</span><br><span class="line">95</span><br><span class="line">96</span><br><span class="line">97</span><br><span class="line">98</span><br><span class="line">99</span><br><span class="line">100</span><br><span class="line">101</span><br><span class="line">102</span><br><span class="line">103</span><br><span class="line">104</span><br><span class="line">105</span><br><span class="line">106</span><br><span class="line">107</span><br><span class="line">108</span><br><span class="line">109</span><br><span class="line">110</span><br><span class="line">111</span><br><span class="line">112</span><br><span class="line">113</span><br><span class="line">114</span><br><span class="line">115</span><br><span class="line">116</span><br><span class="line">117</span><br><span class="line">118</span><br><span class="line">119</span><br><span class="line">120</span><br><span class="line">121</span><br><span class="line">122</span><br><span class="line">123</span><br><span class="line">124</span><br><span class="line">125</span><br><span class="line">126</span><br><span class="line">127</span><br><span class="line">128</span><br><span class="line">129</span><br><span class="line">130</span><br><span class="line">131</span><br><span class="line">132</span><br><span class="line">133</span><br><span class="line">134</span><br><span class="line">135</span><br><span class="line">136</span><br><span class="line">137</span><br><span class="line">138</span><br><span class="line">139</span><br><span class="line">140</span><br><span class="line">141</span><br><span class="line">142</span><br><span class="line">143</span><br><span class="line">144</span><br><span class="line">145</span><br><span class="line">146</span><br><span class="line">147</span><br><span class="line">148</span><br><span class="line">149</span><br><span class="line">150</span><br><span class="line">151</span><br><span class="line">152</span><br><span class="line">153</span><br><span class="line">154</span><br><span class="line">155</span><br><span class="line">156</span><br><span class="line">157</span><br><span class="line">158</span><br><span class="line">159</span><br><span class="line">160</span><br><span class="line">161</span><br><span class="line">162</span><br><span class="line">163</span><br><span class="line">164</span><br><span class="line">165</span><br><span class="line">166</span><br><span class="line">167</span><br><span class="line">168</span><br><span class="line">169</span><br><span class="line">170</span><br><span class="line">171</span><br><span class="line">172</span><br><span class="line">173</span><br><span class="line">174</span><br><span class="line">175</span><br><span class="line">176</span><br><span class="line">177</span><br><span class="line">178</span><br><span class="line">179</span><br><span class="line">180</span><br><span class="line">181</span><br><span class="line">182</span><br><span class="line">183</span><br><span class="line">184</span><br><span class="line">185</span><br><span class="line">186</span><br><span class="line">187</span><br><span class="line">188</span><br><span class="line">189</span><br><span class="line">190</span><br><span class="line">191</span><br><span class="line">192</span><br><span class="line">193</span><br><span class="line">194</span><br><span class="line">195</span><br><span class="line">196</span><br><span class="line">197</span><br><span class="line">198</span><br><span class="line">199</span><br><span class="line">200</span><br><span class="line">201</span><br><span class="line">202</span><br><span class="line">203</span><br><span class="line">204</span><br><span class="line">205</span><br><span class="line">206</span><br><span class="line">207</span><br><span class="line">208</span><br><span class="line">209</span><br><span class="line">210</span><br><span class="line">211</span><br><span class="line">212</span><br><span class="line">213</span><br><span class="line">214</span><br><span class="line">215</span><br><span class="line">216</span><br><span class="line">217</span><br><span class="line">218</span><br><span class="line">219</span><br><span class="line">220</span><br><span class="line">221</span><br><span class="line">222</span><br><span class="line">223</span><br><span class="line">224</span><br><span class="line">225</span><br><span class="line">226</span><br><span class="line">227</span><br><span class="line">228</span><br><span class="line">229</span><br><span class="line">230</span><br><span class="line">231</span><br><span class="line">232</span><br><span class="line">233</span><br><span class="line">234</span><br><span class="line">235</span><br><span class="line">236</span><br><span class="line">237</span><br><span class="line">238</span><br><span class="line">239</span><br><span class="line">240</span><br><span class="line">241</span><br><span class="line">242</span><br><span class="line">243</span><br><span class="line">244</span><br><span class="line">245</span><br><span class="line">246</span><br><span class="line">247</span><br><span class="line">248</span><br><span class="line">249</span><br><span class="line">250</span><br><span class="line">251</span><br><span class="line">252</span><br><span class="line">253</span><br><span class="line">254</span><br><span class="line">255</span><br><span class="line">256</span><br><span class="line">257</span><br><span class="line">258</span><br><span class="line">259</span><br><span class="line">260</span><br><span class="line">261</span><br><span class="line">262</span><br><span class="line">263</span><br><span class="line">264</span><br><span class="line">265</span><br><span class="line">266</span><br><span class="line">267</span><br><span class="line">268</span><br><span class="line">269</span><br><span class="line">270</span><br><span class="line">271</span><br><span class="line">272</span><br><span class="line">273</span><br><span class="line">274</span><br><span class="line">275</span><br><span class="line">276</span><br><span class="line">277</span><br><span class="line">278</span><br><span class="line">279</span><br><span class="line">280</span><br><span class="line">281</span><br><span class="line">282</span><br><span class="line">283</span><br><span class="line">284</span><br><span class="line">285</span><br><span class="line">286</span><br><span class="line">287</span><br><span class="line">288</span><br><span class="line">289</span><br><span class="line">290</span><br><span class="line">291</span><br><span class="line">292</span><br><span class="line">293</span><br><span class="line">294</span><br><span class="line">295</span><br><span class="line">296</span><br><span class="line">297</span><br><span class="line">298</span><br><span class="line">299</span><br><span class="line">300</span><br><span class="line">301</span><br><span class="line">302</span><br><span class="line">303</span><br><span class="line">304</span><br><span class="line">305</span><br><span class="line">306</span><br><span class="line">307</span><br><span class="line">308</span><br><span class="line">309</span><br><span class="line">310</span><br><span class="line">311</span><br><span class="line">312</span><br><span class="line">313</span><br><span class="line">314</span><br><span class="line">315</span><br><span class="line">316</span><br><span class="line">317</span><br><span class="line">318</span><br><span class="line">319</span><br><span class="line">320</span><br><span class="line">321</span><br><span class="line">322</span><br><span class="line">323</span><br><span class="line">324</span><br><span class="line">325</span><br><span class="line">326</span><br><span class="line">327</span><br><span class="line">328</span><br><span class="line">329</span><br><span class="line">330</span><br><span class="line">331</span><br><span class="line">332</span><br><span class="line">333</span><br><span class="line">334</span><br><span class="line">335</span><br><span class="line">336</span><br><span class="line">337</span><br><span class="line">338</span><br><span class="line">339</span><br><span class="line">340</span><br><span class="line">341</span><br><span class="line">342</span><br><span class="line">343</span><br><span class="line">344</span><br><span class="line">345</span><br><span class="line">346</span><br><span class="line">347</span><br><span class="line">348</span><br><span class="line">349</span><br><span class="line">350</span><br><span class="line">351</span><br><span class="line">352</span><br><span class="line">353</span><br><span class="line">354</span><br><span class="line">355</span><br><span class="line">356</span><br><span class="line">357</span><br><span class="line">358</span><br><span class="line">359</span><br><span class="line">360</span><br><span class="line">361</span><br><span class="line">362</span><br><span class="line">363</span><br><span class="line">364</span><br><span class="line">365</span><br><span class="line">366</span><br><span class="line">367</span><br><span class="line">368</span><br><span class="line">369</span><br><span class="line">370</span><br><span class="line">371</span><br><span class="line">372</span><br><span class="line">373</span><br><span class="line">374</span><br><span class="line">375</span><br><span class="line">376</span><br><span class="line">377</span><br><span class="line">378</span><br><span class="line">379</span><br><span class="line">380</span><br><span class="line">381</span><br><span class="line">382</span><br><span class="line">383</span><br><span class="line">384</span><br><span class="line">385</span><br><span class="line">386</span><br><span class="line">387</span><br><span class="line">388</span><br><span class="line">389</span><br><span class="line">390</span><br><span class="line">391</span><br><span class="line">392</span><br><span class="line">393</span><br><span class="line">394</span><br><span class="line">395</span><br><span class="line">396</span><br><span class="line">397</span><br><span class="line">398</span><br><span class="line">399</span><br><span class="line">400</span><br><span class="line">401</span><br><span class="line">402</span><br><span class="line">403</span><br><span class="line">404</span><br><span class="line">405</span><br><span class="line">406</span><br><span class="line">407</span><br><span class="line">408</span><br><span class="line">409</span><br><span class="line">410</span><br><span class="line">411</span><br><span class="line">412</span><br><span class="line">413</span><br><span class="line">414</span><br><span class="line">415</span><br><span class="line">416</span><br><span class="line">417</span><br><span class="line">418</span><br><span class="line">419</span><br><span class="line">420</span><br><span class="line">421</span><br><span class="line">422</span><br><span class="line">423</span><br><span class="line">424</span><br><span class="line">425</span><br><span class="line">426</span><br><span class="line">427</span><br><span class="line">428</span><br><span class="line">429</span><br><span class="line">430</span><br><span class="line">431</span><br><span class="line">432</span><br><span class="line">433</span><br><span class="line">434</span><br><span class="line">435</span><br><span class="line">436</span><br><span class="line">437</span><br><span class="line">438</span><br><span class="line">439</span><br><span class="line">440</span><br><span class="line">441</span><br><span class="line">442</span><br><span class="line">443</span><br><span class="line">444</span><br><span class="line">445</span><br><span class="line">446</span><br><span class="line">447</span><br><span class="line">448</span><br><span class="line">449</span><br><span class="line">450</span><br><span class="line">451</span><br><span class="line">452</span><br><span class="line">453</span><br><span class="line">454</span><br><span class="line">455</span><br><span class="line">456</span><br><span class="line">457</span><br><span class="line">458</span><br><span class="line">459</span><br><span class="line">460</span><br><span class="line">461</span><br><span class="line">462</span><br><span class="line">463</span><br><span class="line">464</span><br><span class="line">465</span><br><span class="line">466</span><br><span class="line">467</span><br><span class="line">468</span><br><span class="line">469</span><br><span class="line">470</span><br><span class="line">471</span><br><span class="line">472</span><br><span class="line">473</span><br><span class="line">474</span><br><span class="line">475</span><br><span class="line">476</span><br><span class="line">477</span><br><span class="line">478</span><br><span class="line">479</span><br><span class="line">480</span><br><span class="line">481</span><br><span class="line">482</span><br><span class="line">483</span><br><span class="line">484</span><br><span class="line">485</span><br><span class="line">486</span><br><span class="line">487</span><br><span class="line">488</span><br><span class="line">489</span><br><span class="line">490</span><br><span class="line">491</span><br><span class="line">492</span><br><span class="line">493</span><br><span class="line">494</span><br><span class="line">495</span><br><span class="line">496</span><br><span class="line">497</span><br><span class="line">498</span><br><span class="line">499</span><br><span class="line">500</span><br><span class="line">501</span><br><span class="line">502</span><br><span class="line">503</span><br><span class="line">504</span><br><span class="line">505</span><br><span class="line">506</span><br><span class="line">507</span><br><span class="line">508</span><br><span class="line">509</span><br><span class="line">510</span><br><span class="line">511</span><br><span class="line">512</span><br><span class="line">513</span><br><span class="line">514</span><br><span class="line">515</span><br><span class="line">516</span><br><span class="line">517</span><br><span class="line">518</span><br><span class="line">519</span><br><span class="line">520</span><br><span class="line">521</span><br><span class="line">522</span><br><span class="line">523</span><br><span class="line">524</span><br><span class="line">525</span><br><span class="line">526</span><br><span class="line">527</span><br><span class="line">528</span><br><span class="line">529</span><br><span class="line">530</span><br><span class="line">531</span><br><span class="line">532</span><br><span class="line">533</span><br><span class="line">534</span><br><span class="line">535</span><br><span class="line">536</span><br><span class="line">537</span><br><span class="line">538</span><br><span class="line">539</span><br><span class="line">540</span><br><span class="line">541</span><br><span class="line">542</span><br><span class="line">543</span><br><span class="line">544</span><br><span class="line">545</span><br><span class="line">546</span><br><span class="line">547</span><br><span class="line">548</span><br><span class="line">549</span><br><span class="line">550</span><br><span class="line">551</span><br><span class="line">552</span><br><span class="line">553</span><br><span class="line">554</span><br><span class="line">555</span><br><span class="line">556</span><br><span class="line">557</span><br><span class="line">558</span><br><span class="line">559</span><br><span class="line">560</span><br><span class="line">561</span><br><span class="line">562</span><br><span class="line">563</span><br><span class="line">564</span><br><span class="line">565</span><br><span class="line">566</span><br><span class="line">567</span><br><span class="line">568</span><br><span class="line">569</span><br><span class="line">570</span><br><span class="line">571</span><br><span class="line">572</span><br><span class="line">573</span><br><span class="line">574</span><br><span class="line">575</span><br><span class="line">576</span><br><span class="line">577</span><br><span class="line">578</span><br><span class="line">579</span><br><span class="line">580</span><br><span class="line">581</span><br><span class="line">582</span><br><span class="line">583</span><br><span class="line">584</span><br><span class="line">585</span><br><span class="line">586</span><br><span class="line">587</span><br><span class="line">588</span><br><span class="line">589</span><br><span class="line">590</span><br><span class="line">591</span><br><span class="line">592</span><br><span class="line">593</span><br><span class="line">594</span><br><span class="line">595</span><br><span class="line">596</span><br><span class="line">597</span><br><span class="line">598</span><br><span class="line">599</span><br><span class="line">600</span><br><span class="line">601</span><br><span class="line">602</span><br><span class="line">603</span><br><span class="line">604</span><br><span class="line">605</span><br><span class="line">606</span><br><span class="line">607</span><br><span class="line">608</span><br><span class="line">609</span><br><span class="line">610</span><br><span class="line">611</span><br><span class="line">612</span><br><span class="line">613</span><br><span class="line">614</span><br><span class="line">615</span><br><span class="line">616</span><br><span class="line">617</span><br><span class="line">618</span><br><span class="line">619</span><br><span class="line">620</span><br><span class="line">621</span><br><span class="line">622</span><br><span class="line">623</span><br><span class="line">624</span><br><span class="line">625</span><br><span class="line">626</span><br><span class="line">627</span><br><span class="line">628</span><br><span class="line">629</span><br><span class="line">630</span><br><span class="line">631</span><br><span class="line">632</span><br><span class="line">633</span><br><span class="line">634</span><br><span class="line">635</span><br><span class="line">636</span><br><span class="line">637</span><br><span class="line">638</span><br><span class="line">639</span><br><span class="line">640</span><br><span class="line">641</span><br><span class="line">642</span><br><span class="line">643</span><br><span class="line">644</span><br><span class="line">645</span><br><span class="line">646</span><br><span class="line">647</span><br><span class="line">648</span><br><span class="line">649</span><br><span class="line">650</span><br><span class="line">651</span><br><span class="line">652</span><br><span class="line">653</span><br><span class="line">654</span><br><span class="line">655</span><br><span class="line">656</span><br><span class="line">657</span><br><span class="line">658</span><br></pre></td><td class="code"><pre><span class="line"><span class="keyword">import</span> pandas <span class="keyword">as</span> pd</span><br><span class="line"><span class="keyword">from</span> sklearn.cluster <span class="keyword">import</span> KMeans</span><br><span class="line"><span class="keyword">from</span> sklearn <span class="keyword">import</span> metrics</span><br><span class="line"><span class="keyword">import</span> re</span><br><span class="line"><span class="keyword">from</span> sklearn.impute <span class="keyword">import</span> SimpleImputer <span class="keyword">as</span> Imputer</span><br><span class="line"><span class="keyword">from</span> numpy <span class="keyword">import</span> random</span><br><span class="line"><span class="keyword">import</span> numpy <span class="keyword">as</span> np</span><br><span class="line"><span class="keyword">import</span> seaborn <span class="keyword">as</span> sb</span><br><span class="line"><span class="keyword">import</span> matplotlib.pyplot <span class="keyword">as</span> plt</span><br><span class="line"></span><br><span class="line"><span class="comment">## proteins data used for PAM50 test</span></span><br><span class="line">pam50_proteins = <span class="string">'PAM50_proteins.csv'</span></span><br><span class="line">pam50 = pd.read_csv(pam50_proteins,header=<span class="number">0</span>)</span><br><span class="line"></span><br><span class="line"><span class="comment">## breast cancer patients' proteomes data</span></span><br><span class="line">proteomes_file = <span class="string">'proteomes_CPTAC_itraq_77_cancer.csv'</span></span><br><span class="line"><span class="comment">## breast cancer patients' clinical data</span></span><br><span class="line">clinical_info = <span class="string">'clinical_data_breast_cancer.csv'</span></span><br><span class="line"></span><br><span class="line">data = pd.read_csv(proteomes_file,header=<span class="number">0</span>,index_col=<span class="number">0</span>)</span><br><span class="line">clinical = pd.read_csv(clinical_info,header=<span class="number">0</span>,index_col=<span class="number">0</span>)</span><br><span class="line"></span><br><span class="line"><span class="comment"># Drop unused information columns</span></span><br><span class="line">data.drop([<span class="string">'gene_symbol'</span>,<span class="string">'gene_name'</span>],axis=<span class="number">1</span>,inplace=<span class="literal">True</span>)</span><br><span class="line"></span><br><span class="line"><span class="comment"># ## Change the protein data sample names to a format matching the clinical data set</span></span><br><span class="line">data.rename(columns=<span class="keyword">lambda</span> x: <span class="string">"TCGA-%s"</span> % (re.split(<span class="string">'[_|-|.]'</span>,x)[<span class="number">0</span>]) <span class="keyword">if</span> bool(re.search(<span class="string">"TCGA"</span>,x)) <span class="keyword">is</span> <span class="literal">True</span> <span class="keyword">else</span> x,inplace=<span class="literal">True</span>)</span><br><span class="line"></span><br><span class="line"><span class="comment">## Transpose data for the clustering algorithm since we want to divide patient samples, not proteins</span></span><br><span class="line">data = data.transpose()</span><br><span class="line"> </span><br><span class="line"><span class="comment"># Drop clinical entries for samples not in our protein data set</span></span><br><span class="line">clinical = clinical.loc[[x <span class="keyword">for</span> x <span class="keyword">in</span> clinical.index.tolist() <span class="keyword">if</span> x <span class="keyword">in</span> data.index],:]</span><br><span class="line"></span><br><span class="line"><span class="comment">## Add clinical meta data to our protein data set, note: all numerical features for analysis start with NP_ or XP_</span></span><br><span class="line">merged = data.merge(clinical,left_index=<span class="literal">True</span>,right_index=<span class="literal">True</span>)</span><br><span class="line"> </span><br><span class="line"><span class="comment">## Change name to make it look nicer in the code!</span></span><br><span class="line">processed = merged</span><br><span class="line"> </span><br><span class="line"><span class="comment">## Numerical data for the algorithm, NP_xx/XP_xx are protein identifiers from RefSeq database</span></span><br><span class="line">processed_numerical = processed.loc[:,[x <span class="keyword">for</span> x <span class="keyword">in</span> processed.columns <span class="keyword">if</span> bool(re.search(<span class="string">"NP_|XP_"</span>,x)) == <span class="literal">True</span>]]</span><br><span class="line"> </span><br><span class="line"><span class="comment">## Select only the PAM50 proteins - known panel of genes used for breast cancer subtype prediction</span></span><br><span class="line">processed_numerical_p50 = processed_numerical.loc[:,processed_numerical.columns.isin(pam50[<span class="string">'RefSeqProteinID'</span>])]</span><br><span class="line">processed_numerical_p50.replace([np.inf, -np.inf], np.nan)</span><br><span class="line"></span><br><span class="line"><span class="comment">## Impute missing values (maybe another method would work better?) </span></span><br><span class="line">imputer = Imputer(missing_values=np.nan, strategy=<span class="string">'mean'</span>)</span><br><span class="line">imputer = imputer.fit(processed_numerical_p50)</span><br><span class="line">processed_numerical_p50 = imputer.transform(processed_numerical_p50)</span><br><span class="line"></span><br><span class="line">k = <span class="number">4</span></span><br><span class="line">clusterer = KMeans(n_clusters=k, n_jobs=<span class="number">4</span>)</span><br><span class="line">clusterer.fit(processed_numerical_p50)</span><br><span class="line">clusterer.labels_</span><br><span class="line"></span><br><span class="line"><span class="function"><span class="keyword">def</span> <span class="title">compare_k_means</span><span class="params">(k_list,data)</span>:</span></span><br><span class="line">    <span class="string">"""</span></span><br><span class="line"><span class="string">        compare_k_means: Run clustering with different k and check the metrics</span></span><br><span class="line"><span class="string"></span></span><br><span class="line"><span class="string">        Args:</span></span><br><span class="line"><span class="string">            k_list (list): different number of k's to determine how many clusters we want</span></span><br><span class="line"><span class="string">            data (dataframe): the data used for clustering</span></span><br><span class="line"><span class="string"></span></span><br><span class="line"><span class="string">        Returns:</span></span><br><span class="line"><span class="string">            None</span></span><br><span class="line"><span class="string">    """</span></span><br><span class="line">    <span class="keyword">for</span> k <span class="keyword">in</span> k_list:</span><br><span class="line">        clusterer = KMeans(n_clusters=k, n_jobs=<span class="number">4</span>)</span><br><span class="line">        clusterer.fit(data)</span><br><span class="line">        <span class="comment"># The higher (up to 1) the better</span></span><br><span class="line">        print(<span class="string">"Silhouette Coefficient for k == %s: %s"</span> % (</span><br><span class="line">        k, round(metrics.silhouette_score(data, clusterer.labels_), <span class="number">4</span>)))</span><br><span class="line">        <span class="comment"># The higher (up to 1) the better</span></span><br><span class="line">        print(<span class="string">"Homogeneity score for k == %s: %s"</span> % (</span><br><span class="line">        k, round(metrics.homogeneity_score(processed[<span class="string">'PAM50 mRNA'</span>], clusterer.labels_), <span class="number">4</span>)))</span><br><span class="line">        print(<span class="string">"------------------------"</span>)</span><br><span class="line"></span><br><span class="line">n_clusters = [<span class="number">2</span>,<span class="number">3</span>,<span class="number">4</span>,<span class="number">5</span>,<span class="number">6</span>,<span class="number">7</span>,<span class="number">8</span>,<span class="number">10</span>,<span class="number">20</span>,<span class="number">79</span>]</span><br><span class="line">compare_k_means(n_clusters,processed_numerical_p50)</span><br><span class="line">processed_numerical_random = processed_numerical.iloc[:,random.choice(range(processed_numerical.shape[<span class="number">1</span>]),<span class="number">43</span>)]</span><br><span class="line">imputer_rnd = imputer.fit(processed_numerical_random)</span><br><span class="line">processed_numerical_random = imputer_rnd.transform(processed_numerical_random)</span><br><span class="line"></span><br><span class="line"><span class="comment">## Use random proteins for comparison</span></span><br><span class="line">print(<span class="string">"====== Now it's the random proteins ======"</span>)</span><br><span class="line">compare_k_means(n_clusters,processed_numerical_random)</span><br><span class="line"><span class="comment">## The scores should be significantly lower than for the PAM50 proteins!</span></span><br><span class="line"></span><br><span class="line"> </span><br><span class="line"><span class="comment">## Visualize data using k==3, show the heatmap of protein expression for the used PAM50 proteins (43 available in our data)</span></span><br><span class="line">clusterer_final = KMeans(n_clusters=<span class="number">3</span>, n_jobs=<span class="number">4</span>)</span><br><span class="line">clusterer_final = clusterer_final.fit(processed_numerical_p50)</span><br><span class="line">processed_p50_plot = pd.DataFrame(processed_numerical_p50)</span><br><span class="line">processed_p50_plot[<span class="string">'KMeans_cluster'</span>] = clusterer_final.labels_</span><br><span class="line">processed_p50_plot.sort_values(<span class="string">'KMeans_cluster'</span>,axis=<span class="number">0</span>,inplace=<span class="literal">True</span>)</span><br><span class="line">fig = plt.figure(figsize = (<span class="number">14</span>,<span class="number">10</span>))</span><br><span class="line">ax = fig.add_subplot(<span class="number">1</span>,<span class="number">1</span>,<span class="number">1</span>)</span><br><span class="line">processed_p50_plot.index.name = <span class="string">'Patient'</span></span><br><span class="line">processed_p50_plot.head()</span><br><span class="line">sb.heatmap(processed_p50_plot.iloc[:,:<span class="number">-1</span>], xticklabels=[], yticklabels=[]) <span class="comment">## The x-axis are the PAM50 proteins we used and the right-most column is the cluster marker</span></span><br><span class="line">ax.set_xlabel(<span class="string">'Protein'</span>, fontsize = <span class="number">20</span>)</span><br><span class="line">ax.set_ylabel(<span class="string">'Patient'</span>, fontsize = <span class="number">20</span>)</span><br><span class="line">ax.set_title(<span class="string">'Protein Expressions for Patients'</span>, fontsize = <span class="number">24</span>)</span><br><span class="line">ax.yaxis.set_tick_params(labelsize=<span class="number">12</span>, rotation=<span class="number">0</span>)</span><br><span class="line">ax.xaxis.set_tick_params(labelsize=<span class="number">12</span>, rotation=<span class="number">0</span>)</span><br><span class="line"></span><br><span class="line"></span><br><span class="line"><span class="keyword">import</span> pandas <span class="keyword">as</span> pd</span><br><span class="line"><span class="keyword">import</span> numpy <span class="keyword">as</span> np</span><br><span class="line"><span class="keyword">import</span> random <span class="keyword">as</span> rd</span><br><span class="line"><span class="keyword">from</span> matplotlib <span class="keyword">import</span> pyplot <span class="keyword">as</span> plt</span><br><span class="line"><span class="keyword">import</span> math</span><br><span class="line"><span class="keyword">from</span> sklearn.preprocessing <span class="keyword">import</span> StandardScaler</span><br><span class="line"><span class="keyword">from</span> sklearn.decomposition <span class="keyword">import</span> PCA <span class="keyword">as</span> sklearnPCA</span><br><span class="line"><span class="keyword">from</span> sklearn <span class="keyword">import</span> preprocessing</span><br><span class="line"><span class="keyword">import</span> sys</span><br><span class="line"><span class="keyword">import</span> seaborn</span><br><span class="line"></span><br><span class="line">s = <span class="number">20</span></span><br><span class="line">wt = [<span class="string">'wt'</span> + str(i) <span class="keyword">for</span> i <span class="keyword">in</span> range(<span class="number">1</span>,s+<span class="number">1</span>)]</span><br><span class="line">ko = [<span class="string">'ko'</span> + str(i) <span class="keyword">for</span> i <span class="keyword">in</span> range(<span class="number">1</span>,s+<span class="number">1</span>)]</span><br><span class="line">genes = [<span class="string">'gene'</span> + str(i) <span class="keyword">for</span> i <span class="keyword">in</span> range(<span class="number">1</span>,<span class="number">101</span>)]</span><br><span class="line"><span class="comment">## Now we create the dataframe</span></span><br><span class="line">data = pd.DataFrame(columns=[*wt, *ko], index=genes)</span><br><span class="line"><span class="keyword">for</span> gene <span class="keyword">in</span> data.index:</span><br><span class="line">    data.loc[gene,[<span class="string">'wt'</span>+str(i) <span class="keyword">for</span> i <span class="keyword">in</span> range(<span class="number">1</span>,s+<span class="number">1</span>)]] = np.random.poisson(lam=rd.randrange(<span class="number">10</span>,<span class="number">1000</span>), size=s)</span><br><span class="line">    data.loc[gene,[<span class="string">'ko'</span>+str(i) <span class="keyword">for</span> i <span class="keyword">in</span> range(<span class="number">1</span>,s+<span class="number">1</span>)]] = np.random.poisson(lam=rd.randrange(<span class="number">10</span>,<span class="number">1000</span>), size=s)</span><br><span class="line">data = data/<span class="number">1000</span></span><br><span class="line">df = data.astype(float)</span><br><span class="line"></span><br><span class="line"><span class="comment">## take a look at the dataset</span></span><br><span class="line">print(df.shape)</span><br><span class="line">df.head()</span><br><span class="line"></span><br><span class="line">df=df.transpose()</span><br><span class="line">x = StandardScaler().fit_transform(df)</span><br><span class="line">sklearn_pca = sklearnPCA(n_components=<span class="number">5</span>)</span><br><span class="line">Y_sklearn = sklearn_pca.fit_transform(x)</span><br><span class="line"></span><br><span class="line">x1=Y_sklearn[:<span class="number">20</span>,:]</span><br><span class="line">x2=Y_sklearn[<span class="number">20</span>:,:]</span><br><span class="line"></span><br><span class="line"><span class="keyword">with</span> plt.style.context(<span class="string">'seaborn-whitegrid'</span>):</span><br><span class="line">    plt.figure(figsize=(<span class="number">10</span>, <span class="number">8</span>))</span><br><span class="line"></span><br><span class="line">    plt.scatter(x1[:, <span class="number">0</span>],</span><br><span class="line">                x1[:, <span class="number">1</span>],</span><br><span class="line">                s=<span class="number">70</span>,</span><br><span class="line">                label=<span class="string">'wt'</span>,</span><br><span class="line">                c=<span class="string">'red'</span>)</span><br><span class="line">    </span><br><span class="line">    plt.scatter(x2[:, <span class="number">0</span>],</span><br><span class="line">                x2[:, <span class="number">1</span>],</span><br><span class="line">                s=<span class="number">70</span>,</span><br><span class="line">                label=<span class="string">'ko'</span>,</span><br><span class="line">                c=<span class="string">'blue'</span>)</span><br><span class="line">        </span><br><span class="line">    plt.xlabel(<span class="string">'Principal Component 1'</span>,fontsize=<span class="number">16</span>)</span><br><span class="line">    plt.ylabel(<span class="string">'Principal Component 2'</span>,fontsize=<span class="number">16</span>)</span><br><span class="line">    plt.xticks(fontsize=<span class="number">14</span>)</span><br><span class="line">    plt.yticks(fontsize=<span class="number">14</span>)</span><br><span class="line">    plt.legend(loc=<span class="string">'best'</span>, fontsize=<span class="number">16</span>)</span><br><span class="line">    plt.tight_layout()</span><br><span class="line">    plt.show()</span><br><span class="line">  </span><br><span class="line"></span><br><span class="line"></span><br><span class="line"><span class="keyword">import</span> matplotlib.pyplot <span class="keyword">as</span> plt</span><br><span class="line"><span class="keyword">import</span> matplotlib.mlab <span class="keyword">as</span> mlab</span><br><span class="line"><span class="keyword">import</span> pandas <span class="keyword">as</span> pd</span><br><span class="line"><span class="keyword">import</span> random <span class="keyword">as</span> rd</span><br><span class="line"><span class="keyword">from</span> sklearn.cluster <span class="keyword">import</span> KMeans</span><br><span class="line"><span class="keyword">from</span> sklearn <span class="keyword">import</span> metrics</span><br><span class="line"><span class="keyword">import</span> re</span><br><span class="line"><span class="keyword">from</span> sklearn.impute <span class="keyword">import</span> SimpleImputer <span class="keyword">as</span> Imputer</span><br><span class="line"><span class="keyword">from</span> numpy <span class="keyword">import</span> random</span><br><span class="line"><span class="keyword">import</span> numpy <span class="keyword">as</span> np</span><br><span class="line"><span class="keyword">import</span> seaborn <span class="keyword">as</span> sns</span><br><span class="line"><span class="keyword">from</span> sklearn <span class="keyword">import</span> preprocessing</span><br><span class="line"><span class="keyword">from</span> sklearn.preprocessing <span class="keyword">import</span> StandardScaler</span><br><span class="line"><span class="keyword">from</span> sklearn.decomposition <span class="keyword">import</span> PCA</span><br><span class="line"></span><br><span class="line"><span class="function"><span class="keyword">def</span> <span class="title">uniqueish_color</span><span class="params">()</span>:</span></span><br><span class="line">    <span class="string">"""There're better ways to generate unique colors, but this isn't awful."""</span></span><br><span class="line">    <span class="keyword">return</span> plt.cm.gist_ncar(np.random.random())</span><br><span class="line">  </span><br><span class="line">colors_ = [<span class="string">'gold'</span>, <span class="string">'magenta'</span>, <span class="string">'orange'</span>, <span class="string">'green'</span>, <span class="string">'crimson'</span>, <span class="string">'grey'</span>, <span class="string">'brown'</span>]</span><br><span class="line"></span><br><span class="line"><span class="comment">### Set name to each file for easier loading</span></span><br><span class="line">dataset_path = <span class="string">"proteomes_CPTAC_itraq_77_cancer.csv"</span></span><br><span class="line">clinical_info = <span class="string">"clinical_data_breast_cancer.csv"</span></span><br><span class="line">pam50_proteins = <span class="string">"PAM50_proteins.csv"</span></span><br><span class="line">BRCA_RNA_dat = <span class="string">"BRCA_RNA_Seq_Samples_Overlap_with_Proteomes_RSEM_est.csv"</span></span><br><span class="line"></span><br><span class="line"><span class="comment">## Load data</span></span><br><span class="line">data = pd.read_csv(dataset_path,header=<span class="number">0</span>,index_col=<span class="number">0</span>)</span><br><span class="line">clinical = pd.read_csv(clinical_info,header=<span class="number">0</span>,index_col=<span class="number">0</span>)<span class="comment">## holds clinical information about each patient/sample</span></span><br><span class="line">pam50 = pd.read_csv(pam50_proteins,header=<span class="number">0</span>)</span><br><span class="line"></span><br><span class="line"><span class="comment">## Drop unused information columns</span></span><br><span class="line">data.drop([<span class="string">'gene_symbol'</span>,<span class="string">'gene_name'</span>],axis=<span class="number">1</span>,inplace=<span class="literal">True</span>)</span><br><span class="line"> </span><br><span class="line"><span class="comment">## Change the protein data sample names to a format matching the clinical data set</span></span><br><span class="line">data.rename(columns=<span class="keyword">lambda</span> x: <span class="string">"TCGA-%s"</span> % (re.split(<span class="string">'[_|-|.]'</span>,x)[<span class="number">0</span>]) <span class="keyword">if</span> bool(re.search(<span class="string">"TCGA"</span>,x)) <span class="keyword">is</span> <span class="literal">True</span> <span class="keyword">else</span> x,inplace=<span class="literal">True</span>)</span><br><span class="line"></span><br><span class="line">data = data.transpose()</span><br><span class="line"> </span><br><span class="line"><span class="comment">## Drop clinical entries for samples not in our protein data set</span></span><br><span class="line">clinical = clinical.loc[[x <span class="keyword">for</span> x <span class="keyword">in</span> clinical.index.tolist() <span class="keyword">if</span> x <span class="keyword">in</span> data.index],:]</span><br><span class="line"></span><br><span class="line"><span class="comment">## Add clinical meta data to our protein data set, note: all numerical features for analysis start with NP_ or XP_</span></span><br><span class="line">merged = data.merge(clinical,left_index=<span class="literal">True</span>,right_index=<span class="literal">True</span>)</span><br><span class="line"> </span><br><span class="line"><span class="comment">## Change name to make it look nicer in the code!</span></span><br><span class="line">processed = merged</span><br><span class="line"></span><br><span class="line"><span class="comment">## Numerical data for the algorithm, NP_xx/XP_xx are protein identifiers from RefSeq database</span></span><br><span class="line">processed_numerical = processed.loc[:,[x <span class="keyword">for</span> x <span class="keyword">in</span> processed.columns <span class="keyword">if</span> bool(re.search(<span class="string">"NP_|XP_"</span>,x)) == <span class="literal">True</span>]]</span><br><span class="line"> </span><br><span class="line"><span class="comment">## Select only the PAM50 proteins - known panel of genes used for breast cancer subtype prediction</span></span><br><span class="line">processed_numerical_p50 = processed_numerical.loc[:,processed_numerical.columns.isin(pam50[<span class="string">'RefSeqProteinID'</span>])]</span><br><span class="line"><span class="comment"># processed_numerical_p50.replace([np.inf, -np.inf], np.nan)</span></span><br><span class="line"></span><br><span class="line"><span class="comment">## Impute missing values (maybe another method would work better?)</span></span><br><span class="line">imputer = Imputer(missing_values=np.nan, strategy=<span class="string">'mean'</span>)</span><br><span class="line">imputer = imputer.fit(processed_numerical_p50)</span><br><span class="line">processed_numerical_p50 = imputer.transform(processed_numerical_p50)</span><br><span class="line"></span><br><span class="line">scaled_data = preprocessing.scale(processed_numerical_p50)</span><br><span class="line"></span><br><span class="line"><span class="comment"># # 2D</span></span><br><span class="line">pca = PCA(n_components=<span class="number">2</span>) <span class="comment"># create a PCA object</span></span><br><span class="line">pca.fit(scaled_data) <span class="comment"># do the math</span></span><br><span class="line">pca_data = pca.transform(scaled_data) <span class="comment"># get PCA coordinates for scaled_data</span></span><br><span class="line">per_var = np.round(pca.explained_variance_ratio_* <span class="number">100</span>, decimals=<span class="number">1</span>)</span><br><span class="line"></span><br><span class="line">principalDf = pd.DataFrame(data = np.array(pca_data), index=processed[<span class="string">'PAM50 mRNA'</span>], columns = [<span class="string">'principal component 1'</span>, <span class="string">'principal component 2'</span>])</span><br><span class="line"></span><br><span class="line"><span class="comment"># # visualization</span></span><br><span class="line">fig = plt.figure(figsize = (<span class="number">8</span>,<span class="number">8</span>))</span><br><span class="line">ax = fig.add_subplot(<span class="number">1</span>,<span class="number">1</span>,<span class="number">1</span>) </span><br><span class="line">ax.set_xlabel(<span class="string">'Principal Component 1 - &#123;0&#125;%'</span>.format(per_var[<span class="number">0</span>]), fontsize = <span class="number">15</span>)</span><br><span class="line">ax.set_ylabel(<span class="string">'Principal Component 2 - &#123;0&#125;%'</span>.format(per_var[<span class="number">1</span>]), fontsize = <span class="number">15</span>)</span><br><span class="line">ax.set_title(<span class="string">'PCA for Proteomics Data - By PAM50'</span>, fontsize = <span class="number">20</span>)</span><br><span class="line">ax.yaxis.set_tick_params(labelsize=<span class="number">14</span>)</span><br><span class="line">ax.xaxis.set_tick_params(labelsize=<span class="number">14</span>)</span><br><span class="line"></span><br><span class="line"><span class="keyword">for</span> name, group <span class="keyword">in</span> principalDf.groupby(principalDf.index):</span><br><span class="line">    group.plot.scatter(x=<span class="string">'principal component 1'</span>, y=<span class="string">'principal component 2'</span>, color=np.random.choice(colors_), s=<span class="number">60</span>, label=name, ax=ax)</span><br><span class="line">ax.legend(fontsize=<span class="number">15</span>)</span><br><span class="line">ax.grid()</span><br><span class="line"></span><br><span class="line"></span><br><span class="line">fig, axes = plt.subplots(<span class="number">1</span>,<span class="number">3</span>,sharey=<span class="literal">True</span>)</span><br><span class="line">fig.set_size_inches((<span class="number">20</span>,<span class="number">7</span>))</span><br><span class="line"></span><br><span class="line"><span class="keyword">for</span> idx <span class="keyword">in</span> range(<span class="number">3</span>, <span class="number">6</span>):</span><br><span class="line">    ax = axes[idx<span class="number">-3</span>]</span><br><span class="line">    clusterer_final = KMeans(n_clusters=idx, n_jobs=<span class="number">4</span>)</span><br><span class="line">clusterer_final = clusterer_final.fit(processed_numerical_p50)</span><br><span class="line"></span><br><span class="line">  principalDf[str(idx)+<span class="string">'_clusters_label'</span>] = clusterer_final.labels_</span><br><span class="line"> ax.yaxis.set_tick_params(labelsize=<span class="number">14</span>)</span><br><span class="line">    ax.xaxis.set_tick_params(labelsize=<span class="number">14</span>)</span><br><span class="line"></span><br><span class="line">  <span class="keyword">for</span> name, group <span class="keyword">in</span> principalDf.groupby(principalDf[str(idx)+<span class="string">'_clusters_label'</span>]):</span><br><span class="line">        group.plot.scatter(x=<span class="string">'principal component 1'</span>, y=<span class="string">'principal component 2'</span>, color=np.random.choice(colors_), s=<span class="number">60</span>, label=<span class="string">'Cluster '</span>+str(name+<span class="number">1</span>), ax=ax)</span><br><span class="line">    </span><br><span class="line">    <span class="keyword">if</span> (idx<span class="number">-3</span> == <span class="number">1</span>):</span><br><span class="line">        ax.set_xlabel(<span class="string">'Principal Component 1 - &#123;0&#125;%'</span>.format(per_var[<span class="number">0</span>]), fontsize = <span class="number">15</span>)</span><br><span class="line">        ax.set_title(<span class="string">'PCA for All Proteomics Data - By K-Means Clustering'</span>, fontsize = <span class="number">20</span>)</span><br><span class="line">    <span class="keyword">else</span>:</span><br><span class="line">        ax.set_xlabel(<span class="string">''</span>)</span><br><span class="line">        </span><br><span class="line">    <span class="keyword">if</span> (idx<span class="number">-3</span> == <span class="number">0</span>):</span><br><span class="line">        ax.set_ylabel(<span class="string">'Principal Component 2 - &#123;0&#125;%'</span>.format(per_var[<span class="number">1</span>]), fontsize = <span class="number">15</span>)</span><br><span class="line">        </span><br><span class="line">    ax.legend(fontsize=<span class="number">14</span>)</span><br><span class="line">    ax.grid()</span><br><span class="line"></span><br><span class="line">plt.tight_layout()</span><br><span class="line">plt.subplots_adjust(top=<span class="number">0.88</span>,wspace = <span class="number">0.03</span>,hspace=<span class="number">0.12</span>)</span><br><span class="line"></span><br><span class="line"></span><br><span class="line">processed_numerical_random = processed_numerical.iloc[:,random.choice(range(processed_numerical.shape[<span class="number">1</span>]),<span class="number">60</span>)]</span><br><span class="line">imputer_rnd = imputer.fit(processed_numerical_random)</span><br><span class="line">processed_numerical_random = imputer_rnd.transform(processed_numerical_random)</span><br><span class="line"></span><br><span class="line">fig, axes = plt.subplots(<span class="number">1</span>,<span class="number">3</span>,sharey=<span class="literal">True</span>)</span><br><span class="line">fig.set_size_inches((<span class="number">20</span>,<span class="number">7</span>))</span><br><span class="line"></span><br><span class="line"><span class="keyword">for</span> idx <span class="keyword">in</span> range(<span class="number">3</span>, <span class="number">6</span>):</span><br><span class="line">    ax = axes[idx<span class="number">-3</span>]</span><br><span class="line">    clusterer_final = KMeans(n_clusters=idx, n_jobs=<span class="number">4</span>)</span><br><span class="line">    clusterer_final = clusterer_final.fit(processed_numerical_random)</span><br><span class="line"></span><br><span class="line">  principalDf[str(idx)+<span class="string">'_clusters_label'</span>] = clusterer_final.labels_</span><br><span class="line"> ax.yaxis.set_tick_params(labelsize=<span class="number">14</span>)</span><br><span class="line">    ax.xaxis.set_tick_params(labelsize=<span class="number">14</span>)</span><br><span class="line">  <span class="comment"># principalDf.plot.scatter(x='principal component 1', y='principal component 2', label=principalDf.index, ax=ax);</span></span><br><span class="line">    <span class="keyword">for</span> name, group <span class="keyword">in</span> principalDf.groupby(principalDf[str(idx)+<span class="string">'_clusters_label'</span>]):</span><br><span class="line">        group.plot.scatter(x=<span class="string">'principal component 1'</span>, y=<span class="string">'principal component 2'</span>, color=np.random.choice(colors_), s=<span class="number">60</span>, label=<span class="string">'Cluster '</span>+str(name+<span class="number">1</span>), ax=ax)</span><br><span class="line">    </span><br><span class="line">    <span class="keyword">if</span> (idx<span class="number">-3</span> == <span class="number">1</span>):</span><br><span class="line">        ax.set_xlabel(<span class="string">'Principal Component 1 - &#123;0&#125;%'</span>.format(per_var[<span class="number">0</span>]), fontsize = <span class="number">15</span>)</span><br><span class="line">        ax.set_title(<span class="string">'PCA for Random Proteomics Data - By K-Means Clustering'</span>, fontsize = <span class="number">20</span>)</span><br><span class="line">    <span class="keyword">else</span>:</span><br><span class="line">        ax.set_xlabel(<span class="string">''</span>)</span><br><span class="line">        </span><br><span class="line">    <span class="keyword">if</span> (idx<span class="number">-3</span> == <span class="number">0</span>):</span><br><span class="line">        ax.set_ylabel(<span class="string">'Principal Component 2 - &#123;0&#125;%'</span>.format(per_var[<span class="number">1</span>]), fontsize = <span class="number">15</span>)</span><br><span class="line">        </span><br><span class="line">    ax.legend(fontsize=<span class="number">14</span>)</span><br><span class="line">    ax.grid()</span><br><span class="line"></span><br><span class="line">plt.tight_layout()</span><br><span class="line">plt.subplots_adjust(top=<span class="number">0.88</span>,wspace = <span class="number">0.03</span>,hspace=<span class="number">0.12</span>)</span><br><span class="line"></span><br><span class="line">BRCA_RNA_df = pd.read_csv(<span class="string">"BRCA_RNA_Seq_Samples_Overlap_with_Proteomes_RSEM_est.csv"</span>, index_col=<span class="number">0</span>)</span><br><span class="line">BRCA_RNA_df = BRCA_RNA_df * <span class="number">1e6</span></span><br><span class="line"></span><br><span class="line">BRCA_RNA_df=BRCA_RNA_df.loc[BRCA_RNA_df.index!=np.nan,:]</span><br><span class="line">idx = BRCA_RNA_df.index</span><br><span class="line"><span class="comment"># BRCA_RNA_df.index = list(map(lambda x: x.split('|')[0].strip(), idx))</span></span><br><span class="line"></span><br><span class="line"><span class="comment">### Filter out non-expressed genes</span></span><br><span class="line">BRCA_RNA_df = BRCA_RNA_df.loc[BRCA_RNA_df.index != <span class="string">'?'</span>, :]</span><br><span class="line">BRCA_RNA_df = BRCA_RNA_df.loc[BRCA_RNA_df.sum(axis=<span class="number">1</span>) &gt; <span class="number">0</span>, :]</span><br><span class="line"></span><br><span class="line"><span class="comment">## Filter out lowly expressed genes</span></span><br><span class="line">mask_low_vals = (BRCA_RNA_df &gt; <span class="number">0.3</span>).sum(axis=<span class="number">1</span>) &gt; <span class="number">2</span></span><br><span class="line">BRCA_RNA_df = BRCA_RNA_df.loc[mask_low_vals, :]</span><br><span class="line"></span><br><span class="line">scaled_data = BRCA_RNA_df.T</span><br><span class="line"><span class="comment"># scaled_data = preprocessing.scale(scaled_data)</span></span><br><span class="line"></span><br><span class="line"><span class="comment"># # 2D</span></span><br><span class="line">pca = PCA(n_components=<span class="number">2</span>) <span class="comment"># create a PCA object</span></span><br><span class="line">pca.fit(scaled_data) <span class="comment"># do the math</span></span><br><span class="line">pca_data = pca.transform(scaled_data) <span class="comment"># get PCA coordinates for scaled_data</span></span><br><span class="line">per_var = np.round(pca.explained_variance_ratio_* <span class="number">100</span>, decimals=<span class="number">1</span>) <span class="comment"># get the variance explained by each component</span></span><br><span class="line"></span><br><span class="line">principalDf = pd.DataFrame(data = np.array(pca_data), index=BRCA_RNA_df.columns, columns = [<span class="string">'principal component 1'</span>, <span class="string">'principal component 2'</span>])</span><br><span class="line">principalDf = principalDf.merge(processed, left_index=<span class="literal">True</span>, right_index=<span class="literal">True</span>)</span><br><span class="line">principalDf = principalDf[[<span class="string">'principal component 1'</span>, <span class="string">'principal component 2'</span>, <span class="string">'PAM50 mRNA'</span>]].drop_duplicates()</span><br><span class="line"></span><br><span class="line"><span class="comment"># # visualization</span></span><br><span class="line">fig = plt.figure(figsize = (<span class="number">10</span>,<span class="number">10</span>))</span><br><span class="line">ax = fig.add_subplot(<span class="number">1</span>,<span class="number">1</span>,<span class="number">1</span>) </span><br><span class="line"></span><br><span class="line"><span class="comment"># principalDf.plot.scatter(x='principal component 1', y='principal component 2', label=principalDf.index, ax=ax);</span></span><br><span class="line"><span class="keyword">for</span> name, group <span class="keyword">in</span> principalDf.groupby(principalDf[<span class="string">'PAM50 mRNA'</span>]):</span><br><span class="line">    group.plot.scatter(x=<span class="string">'principal component 1'</span>, y=<span class="string">'principal component 2'</span>, color=np.random.choice(colors_), s=<span class="number">60</span>, label=name, ax=ax)</span><br><span class="line"></span><br><span class="line">ax.set_xlabel(<span class="string">'Principal Component 1 - &#123;0&#125;%'</span>.format(per_var[<span class="number">0</span>]), fontsize = <span class="number">15</span>)</span><br><span class="line">ax.set_ylabel(<span class="string">'Principal Component 2 - &#123;0&#125;%'</span>.format(per_var[<span class="number">1</span>]), fontsize = <span class="number">15</span>)</span><br><span class="line">ax.set_title(<span class="string">'PCA for RNA Data - By PAM50'</span>, fontsize = <span class="number">20</span>)</span><br><span class="line">ax.yaxis.set_tick_params(labelsize=<span class="number">14</span>)</span><br><span class="line">ax.xaxis.set_tick_params(labelsize=<span class="number">14</span>)</span><br><span class="line"><span class="comment"># ax.set_yscale('log', basey=10)</span></span><br><span class="line"><span class="comment"># ax.set_xscale('log', basex=10)</span></span><br><span class="line"></span><br><span class="line">ax.legend(fontsize=<span class="number">15</span>)</span><br><span class="line">ax.grid()</span><br><span class="line"></span><br><span class="line"><span class="comment">### Filter out the outliers detected by PCA</span></span><br><span class="line">principalDf_ = principalDf[principalDf[<span class="string">'principal component 1'</span>] &lt; <span class="number">40000</span>]</span><br><span class="line">principalDf_ = principalDf_[principalDf_[<span class="string">'principal component 2'</span>] &lt; <span class="number">40000</span>]</span><br><span class="line"></span><br><span class="line"><span class="comment"># # visualization</span></span><br><span class="line">fig = plt.figure(figsize = (<span class="number">10</span>,<span class="number">10</span>))</span><br><span class="line">ax = fig.add_subplot(<span class="number">1</span>,<span class="number">1</span>,<span class="number">1</span>) </span><br><span class="line"></span><br><span class="line"><span class="comment"># principalDf.plot.scatter(x='principal component 1', y='principal component 2', label=principalDf.index, ax=ax);</span></span><br><span class="line"><span class="keyword">for</span> name, group <span class="keyword">in</span> principalDf_.groupby(principalDf_[<span class="string">'PAM50 mRNA'</span>]):</span><br><span class="line">    group.plot.scatter(x=<span class="string">'principal component 1'</span>, y=<span class="string">'principal component 2'</span>, color=np.random.choice(colors_), s=<span class="number">60</span>, label=name, ax=ax)</span><br><span class="line"></span><br><span class="line">ax.set_xlabel(<span class="string">'Principal Component 1 - &#123;0&#125;%'</span>.format(per_var[<span class="number">0</span>]), fontsize = <span class="number">15</span>)</span><br><span class="line">ax.set_ylabel(<span class="string">'Principal Component 2 - &#123;0&#125;%'</span>.format(per_var[<span class="number">1</span>]), fontsize = <span class="number">15</span>)</span><br><span class="line">ax.set_title(<span class="string">'PCA for RNA Data (No Outliers) - By PAM50'</span>, fontsize = <span class="number">20</span>)</span><br><span class="line">ax.yaxis.set_tick_params(labelsize=<span class="number">14</span>)</span><br><span class="line">ax.xaxis.set_tick_params(labelsize=<span class="number">14</span>)</span><br><span class="line"><span class="comment"># ax.set_yscale('log', basey=10)</span></span><br><span class="line"><span class="comment"># ax.set_xscale('log', basex=10)</span></span><br><span class="line"></span><br><span class="line">ax.legend(fontsize=<span class="number">15</span>)</span><br><span class="line">ax.grid()</span><br><span class="line"></span><br><span class="line">rows_to_drop = principalDf.index[principalDf[<span class="string">'principal component 1'</span>] &gt; <span class="number">40000</span>].tolist() + \</span><br><span class="line">               principalDf.index[principalDf[<span class="string">'principal component 2'</span>] &gt; <span class="number">40000</span>].tolist()</span><br><span class="line"></span><br><span class="line">BRCA_RNA_df_no_outliers = BRCA_RNA_df.drop(rows_to_drop, axis=<span class="number">1</span>, inplace=<span class="literal">False</span>)</span><br><span class="line">scaled_data_ = BRCA_RNA_df_no_outliers.T</span><br><span class="line"><span class="comment"># scaled_data = preprocessing.scale(scaled_data)</span></span><br><span class="line"></span><br><span class="line"><span class="comment"># # 2D</span></span><br><span class="line">pca = PCA(n_components=<span class="number">2</span>) <span class="comment"># create a PCA object</span></span><br><span class="line">pca.fit(scaled_data_) <span class="comment"># do the math</span></span><br><span class="line">pca_data_ = pca.transform(scaled_data_) <span class="comment"># get PCA coordinates for scaled_data</span></span><br><span class="line">per_var_ = np.round(pca.explained_variance_ratio_* <span class="number">100</span>, decimals=<span class="number">1</span>) <span class="comment"># get the variance explained by each component</span></span><br><span class="line"></span><br><span class="line">principalDf_ = pd.DataFrame(data = np.array(pca_data_), index=BRCA_RNA_df_no_outliers.columns, columns = [<span class="string">'principal component 1'</span>, <span class="string">'principal component 2'</span>])</span><br><span class="line">principalDf_ = principalDf_.merge(processed, left_index=<span class="literal">True</span>, right_index=<span class="literal">True</span>)</span><br><span class="line">principalDf_ = principalDf_[[<span class="string">'principal component 1'</span>, <span class="string">'principal component 2'</span>, <span class="string">'PAM50 mRNA'</span>]].drop_duplicates()</span><br><span class="line"></span><br><span class="line"></span><br><span class="line"><span class="comment"># # visualization</span></span><br><span class="line">fig = plt.figure(figsize = (<span class="number">10</span>,<span class="number">10</span>))</span><br><span class="line">ax = fig.add_subplot(<span class="number">1</span>,<span class="number">1</span>,<span class="number">1</span>)</span><br><span class="line"></span><br><span class="line"><span class="comment"># principalDf.plot.scatter(x='principal component 1', y='principal component 2', label=principalDf.index, ax=ax);</span></span><br><span class="line"><span class="keyword">for</span> name, group <span class="keyword">in</span> principalDf_.groupby(principalDf_[<span class="string">'PAM50 mRNA'</span>]):</span><br><span class="line">    group.plot.scatter(x=<span class="string">'principal component 1'</span>, y=<span class="string">'principal component 2'</span>, color=np.random.choice(colors_), s=<span class="number">60</span>, label=name, ax=ax)</span><br><span class="line"></span><br><span class="line">ax.set_xlabel(<span class="string">'Principal Component 1 - &#123;0&#125;%'</span>.format(per_var[<span class="number">0</span>]), fontsize = <span class="number">15</span>)</span><br><span class="line">ax.set_ylabel(<span class="string">'Principal Component 2 - &#123;0&#125;%'</span>.format(per_var[<span class="number">1</span>]), fontsize = <span class="number">15</span>)</span><br><span class="line">ax.set_title(<span class="string">'PCA for RNA Data (Refit without outliers) - By PAM50'</span>, fontsize = <span class="number">20</span>)</span><br><span class="line">ax.yaxis.set_tick_params(labelsize=<span class="number">14</span>)</span><br><span class="line">ax.xaxis.set_tick_params(labelsize=<span class="number">14</span>)</span><br><span class="line"><span class="comment"># ax.set_yscale('log', basey=10)</span></span><br><span class="line"><span class="comment"># ax.set_xscale('log', basex=10)</span></span><br><span class="line"></span><br><span class="line">ax.legend(fontsize=<span class="number">15</span>)</span><br><span class="line">ax.grid()</span><br><span class="line"></span><br><span class="line"><span class="keyword">import</span> warnings</span><br><span class="line"><span class="keyword">from</span> scipy.stats <span class="keyword">import</span> chi2</span><br><span class="line"><span class="keyword">from</span> scipy.stats.mstats <span class="keyword">import</span> zscore</span><br><span class="line">warnings.filterwarnings(<span class="string">"ignore"</span>, category=DeprecationWarning) </span><br><span class="line">warnings.filterwarnings(<span class="string">"ignore"</span>, category=RuntimeWarning) </span><br><span class="line"></span><br><span class="line"><span class="function"><span class="keyword">def</span> <span class="title">chdir</span><span class="params">(data, sampleclass, genes, gamma=<span class="number">1.</span>, sort=True, calculate_sig=False, nnull=<span class="number">10</span>, sig_only=False, norm_vector=True)</span>:</span></span><br><span class="line">	<span class="string">"""</span></span><br><span class="line"><span class="string">	Calculate the characteristic direction for a gene expression dataset</span></span><br><span class="line"><span class="string">	</span></span><br><span class="line"><span class="string">	Input:</span></span><br><span class="line"><span class="string">		data: numpy.array, is the data matrix of gene expression where rows correspond to genes and columns correspond to samples</span></span><br><span class="line"><span class="string">		sampleclass: list or numpy.array, labels of the samples, it has to be consist of 0, 1 and 2, with 0 being columns to be excluded, 1 being control and 2 being perturbation</span></span><br><span class="line"><span class="string">				example: sampleclass = [1,1,1,2,2,2]</span></span><br><span class="line"><span class="string">		genes: list or numpy.array, row labels for genes </span></span><br><span class="line"><span class="string">		gamma: float, regulaized term. A parameter that smooths the covariance matrix and reduces potential noise in the dataset</span></span><br><span class="line"><span class="string">		sort: bool, whether to sort the output by the absolute value of chdir</span></span><br><span class="line"><span class="string">		calculate_sig: bool, whether to calculate the significance of characteristic directions</span></span><br><span class="line"><span class="string">		nnull: int, number of null characteristic directions to calculate for significance</span></span><br><span class="line"><span class="string">		sig_only: bool, whether to return only significant genes; active only when calculate_sig is True</span></span><br><span class="line"><span class="string">		norm_vector: bool, whether to return a characteristic direction vector normalized to unit vector</span></span><br><span class="line"><span class="string">	Output:</span></span><br><span class="line"><span class="string">		A list of tuples sorted by the absolute value in descending order characteristic directions of genes.</span></span><br><span class="line"><span class="string">			If calculate_sig is set to True, each tuple contains a third element which is the ratio of characteristic directions to null ChDir</span></span><br><span class="line"><span class="string">	"""</span></span><br><span class="line">	</span><br><span class="line">	<span class="comment">## check input</span></span><br><span class="line">	data.astype(float)</span><br><span class="line">	<span class="comment"># sampleclass = np.array(map(int, sampleclass))</span></span><br><span class="line">	<span class="comment"># masks</span></span><br><span class="line">	m_non0 = sampleclass != <span class="number">0</span></span><br><span class="line">	m1 = sampleclass[m_non0] == <span class="number">1</span></span><br><span class="line">	m2 = sampleclass[m_non0] == <span class="number">2</span></span><br><span class="line"></span><br><span class="line">	<span class="comment"># if type(gamma) not in [float, int]:</span></span><br><span class="line">	<span class="comment"># 	raise ValueError("gamma has to be a numeric number")</span></span><br><span class="line">	<span class="comment"># if set(sampleclass) != set([1,2]) and set(sampleclass) != set([0,1,2]):</span></span><br><span class="line">	<span class="comment"># 	raise ValueError("sampleclass has to be a list whose elements are in only 0, 1 or 2")</span></span><br><span class="line">	<span class="comment"># if m1.sum()&lt;2 or m2.sum()&lt;2:</span></span><br><span class="line">	<span class="comment"># 	raise ValueError("Too few samples to calculate characteristic directions")</span></span><br><span class="line">	<span class="comment"># if len(genes) != data.shape[0]:</span></span><br><span class="line">	<span class="comment"># 	raise ValueError("Number of genes does not match the demension of the expression matrix")</span></span><br><span class="line"></span><br><span class="line">	<span class="comment">## normalize data</span></span><br><span class="line">	data = data[:, m_non0]</span><br><span class="line">	data = zscore(data) <span class="comment"># standardize for each genes across samples</span></span><br><span class="line"></span><br><span class="line">	<span class="comment">## start to compute</span></span><br><span class="line">	n1 = m1.sum() <span class="comment"># number of controls</span></span><br><span class="line">	n2 = m2.sum() <span class="comment"># number of experiments</span></span><br><span class="line"></span><br><span class="line">	<span class="comment">## the difference between experiment mean vector and control mean vector.</span></span><br><span class="line">	meanvec = data[:,m2].mean(axis=<span class="number">1</span>) - data[:,m1].mean(axis=<span class="number">1</span>) </span><br><span class="line"></span><br><span class="line">	<span class="comment">## initialize the pca object</span></span><br><span class="line">	pca = PCA(n_components=<span class="literal">None</span>)</span><br><span class="line">	pca.fit(data.T)</span><br><span class="line"></span><br><span class="line">	<span class="comment">## compute the number of PCs to keep</span></span><br><span class="line">	cumsum = pca.explained_variance_ratio_ <span class="comment"># explained variance of each PC</span></span><br><span class="line">	keepPC = len(cumsum[cumsum &gt; <span class="number">0.001</span>]) <span class="comment"># number of PCs to keep</span></span><br><span class="line"></span><br><span class="line">	v = pca.components_[<span class="number">0</span>:keepPC].T <span class="comment"># rotated data </span></span><br><span class="line">	r = pca.transform(data.T)[:,<span class="number">0</span>:keepPC] <span class="comment"># transformed data</span></span><br><span class="line"></span><br><span class="line">	dd = ( np.dot(r[m1].T,r[m1]) + np.dot(r[m2].T,r[m2]) ) / float(n1+n2<span class="number">-2</span>) <span class="comment"># covariance</span></span><br><span class="line">	sigma = np.mean(np.diag(dd)) <span class="comment"># the scalar covariance</span></span><br><span class="line"></span><br><span class="line">	shrunkMats = np.linalg.inv(gamma*dd + sigma*(<span class="number">1</span>-gamma)*np.eye(keepPC))</span><br><span class="line"></span><br><span class="line">	b = np.dot(v, np.dot(np.dot(v.T, meanvec), shrunkMats))</span><br><span class="line"></span><br><span class="line">	<span class="keyword">if</span> norm_vector:</span><br><span class="line">		b /= np.linalg.norm(b) <span class="comment"># normalize b to unit vector</span></span><br><span class="line"></span><br><span class="line">	grouped = zip([abs(item) <span class="keyword">for</span> item <span class="keyword">in</span> b],b,genes)</span><br><span class="line">	<span class="keyword">if</span> sort:</span><br><span class="line">		grouped = sorted(grouped,key=<span class="keyword">lambda</span> x: x[<span class="number">0</span>], reverse=<span class="literal">True</span>)</span><br><span class="line"></span><br><span class="line"></span><br><span class="line">	<span class="keyword">if</span> <span class="keyword">not</span> calculate_sig: <span class="comment"># return sorted b and genes.</span></span><br><span class="line">		res = [(item[<span class="number">1</span>],item[<span class="number">2</span>]) <span class="keyword">for</span> item <span class="keyword">in</span> grouped]</span><br><span class="line">		<span class="keyword">return</span> res</span><br><span class="line">	<span class="keyword">else</span>: <span class="comment"># generate a null distribution of chdirs</span></span><br><span class="line">		nu = n1 + n2 - <span class="number">2</span></span><br><span class="line">		y1 = np.random.multivariate_normal(np.zeros(keepPC), dd, nnull).T * np.sqrt(nu / chi2.rvs(nu,size=nnull))</span><br><span class="line">		y2 = np.random.multivariate_normal(np.zeros(keepPC), dd, nnull).T * np.sqrt(nu / chi2.rvs(nu,size=nnull))</span><br><span class="line">		y = y2 - y1 <span class="comment">## y is the null of v</span></span><br><span class="line"></span><br><span class="line">		nullchdirs = []</span><br><span class="line">		<span class="keyword">for</span> col <span class="keyword">in</span> y.T:</span><br><span class="line">			bn = np.dot(np.dot(np.dot(v,shrunkMats), v.T), np.dot(col,v.T))</span><br><span class="line">			bn /= np.linalg.norm(bn)</span><br><span class="line">			bn = bn ** <span class="number">2</span></span><br><span class="line">			bn.sort()</span><br><span class="line">			bn = bn[::<span class="number">-1</span>] <span class="comment">## sort in decending order</span></span><br><span class="line">			nullchdirs.append(bn)</span><br><span class="line"></span><br><span class="line">		nullchdirs = np.array(nullchdirs).T</span><br><span class="line">		nullchdirs = nullchdirs.mean(axis=<span class="number">1</span>)</span><br><span class="line">		b_s = b ** <span class="number">2</span> </span><br><span class="line">		b_s.sort()</span><br><span class="line">		b_s = b_s[::<span class="number">-1</span>] <span class="comment"># sorted b in decending order</span></span><br><span class="line">		relerr = b_s / nullchdirs <span class="comment">## relative error</span></span><br><span class="line">		<span class="comment"># ratio_to_null</span></span><br><span class="line">		ratios = np.cumsum(relerr)/np.sum(relerr)- np.linspace(<span class="number">1.</span>/len(meanvec),<span class="number">1</span>,len(meanvec))</span><br><span class="line">		res = [(item[<span class="number">1</span>],item[<span class="number">2</span>], ratio) <span class="keyword">for</span> item, ratio <span class="keyword">in</span> zip(grouped, ratios)] </span><br><span class="line">		<span class="keyword">print</span> (<span class="string">'Number of significant genes: %s'</span>%(np.argmax(ratios)+<span class="number">1</span>))</span><br><span class="line">		<span class="keyword">if</span> sig_only:</span><br><span class="line">			<span class="keyword">return</span> res[<span class="number">0</span>:np.argmax(ratios)+<span class="number">1</span>]</span><br><span class="line">		<span class="keyword">else</span>:</span><br><span class="line">			<span class="keyword">return</span> res</span><br><span class="line"></span><br><span class="line"></span><br><span class="line"><span class="function"><span class="keyword">def</span> <span class="title">paea</span><span class="params">(chdir, gmtline, case_sensitive=False)</span>:</span></span><br><span class="line">	<span class="string">"""</span></span><br><span class="line"><span class="string">	Perform principal angle enrichment analysis (PAEA)</span></span><br><span class="line"><span class="string">	Input:</span></span><br><span class="line"><span class="string">		chdir, list of tuples: A characteristic direction returned from chdir function</span></span><br><span class="line"><span class="string">		gmtline: A list of genes from a gene set </span></span><br><span class="line"><span class="string">	Output:</span></span><br><span class="line"><span class="string">		a list of tuples in the format of:</span></span><br><span class="line"><span class="string">			the principal angle, the p value</span></span><br><span class="line"><span class="string">	"""</span></span><br><span class="line">	<span class="keyword">if</span> len(chdir[<span class="number">0</span>]) == <span class="number">3</span>: <span class="comment">## output from which calculate_sig is enabled</span></span><br><span class="line">		chdir = [(item[<span class="number">0</span>],item[<span class="number">1</span>]) <span class="keyword">for</span> item <span class="keyword">in</span> chdir]</span><br><span class="line">	<span class="keyword">if</span> <span class="keyword">not</span> case_sensitive:</span><br><span class="line">		genes_measured = [gene.upper() <span class="keyword">for</span> b, gene <span class="keyword">in</span> chdir]</span><br><span class="line">		gmtline = [gene.upper() <span class="keyword">for</span> gene <span class="keyword">in</span> gmtline]</span><br><span class="line">	<span class="keyword">else</span>: <span class="comment"># case sensitive</span></span><br><span class="line">		genes_measured = [gene <span class="keyword">for</span> b, gene <span class="keyword">in</span> chdir]</span><br><span class="line"></span><br><span class="line">	<span class="keyword">if</span> len(set(genes_measured)) != len(chdir):</span><br><span class="line">		<span class="keyword">raise</span> ValueError(<span class="string">'There are duplicated genes in the input genes'</span>)</span><br><span class="line">	</span><br><span class="line">	mask = np.in1d(genes_measured, gmtline) <span class="comment"># gpos in the R script</span></span><br><span class="line">	mm = np.where(mask==<span class="literal">True</span>)[<span class="number">0</span>]</span><br><span class="line">	m = mask.sum() <span class="comment"># number of overlaping genes</span></span><br><span class="line">	n = len(genes_measured)</span><br><span class="line"></span><br><span class="line">	<span class="keyword">if</span> m &gt; <span class="number">1</span> <span class="keyword">and</span> m &lt; n: <span class="comment"># if there is overlap between gene set and genes in chdir</span></span><br><span class="line">		gsa = np.zeros((n, m)) <span class="comment"># Qc in the paper</span></span><br><span class="line">		<span class="keyword">for</span> i <span class="keyword">in</span> range(m):</span><br><span class="line">			gsa[mm[i], i] = <span class="number">1.</span></span><br><span class="line"></span><br><span class="line">		qb = np.array([b <span class="keyword">for</span> b,gene <span class="keyword">in</span> chdir])</span><br><span class="line">		qbqc = np.matrix( np.dot(qb, gsa) ) <span class="comment"># Qb.T Qc in paper</span></span><br><span class="line">		principal_angle = np.linalg.svd(qbqc,compute_uv=<span class="literal">False</span>)[<span class="number">0</span>]</span><br><span class="line">		theta = np.arccos(principal_angle)</span><br><span class="line"></span><br><span class="line">		<span class="comment"># calculation of the null-distribution of principal angles</span></span><br><span class="line">		m = float(m)</span><br><span class="line">		n = float(n)</span><br><span class="line">		pac = <span class="keyword">lambda</span> theta: <span class="number">2.</span>*(<span class="number">1.</span>/np.sqrt(<span class="number">2</span>*np.pi))*np.exp((n/<span class="number">2.</span>)*np.log(n/(n - m))+(m/<span class="number">2.</span>)*np.log((n - m)/m)+(<span class="number">1</span>/<span class="number">2.</span>)*np.log(m/(<span class="number">2.</span>*n)*(n - m))+(n-m<span class="number">-1</span>)*np.log(np.sin(theta))+(m<span class="number">-1</span>)*np.log(np.cos(theta)))</span><br><span class="line">		integration_range = np.linspace(<span class="number">0</span>, theta, num=<span class="number">10000</span>, endpoint=<span class="literal">True</span>) <span class="comment">## num seems to matter a lot</span></span><br><span class="line">		p_val = np.trapz(pac(integration_range), integration_range)</span><br><span class="line">		<span class="keyword">if</span> p_val &gt; <span class="number">1.</span>:</span><br><span class="line">			p_val = <span class="number">1.</span></span><br><span class="line">	<span class="keyword">else</span>:</span><br><span class="line">		principal_angle = <span class="number">0.</span></span><br><span class="line">		p_val = <span class="number">1.</span></span><br><span class="line"></span><br><span class="line">	<span class="keyword">return</span> principal_angle, p_val</span><br><span class="line"></span><br><span class="line"></span><br><span class="line"><span class="function"><span class="keyword">def</span> <span class="title">paea_wrapper</span><span class="params">(chdir, gmt_fn, case_sensitive=False, sort=True)</span>:</span></span><br><span class="line">	<span class="string">"""</span></span><br><span class="line"><span class="string">	A wrapper function for PAEA gene-set enrichment analysis</span></span><br><span class="line"><span class="string"></span></span><br><span class="line"><span class="string">	Input:</span></span><br><span class="line"><span class="string">		chdir: characteristic directions computed by chdir function</span></span><br><span class="line"><span class="string">		gmt_fn: file name of a gene-set library in GMT format</span></span><br><span class="line"><span class="string">		case_sensitive: whether gene symbols should be considered as case_sensitive</span></span><br><span class="line"><span class="string">	Output:</span></span><br><span class="line"><span class="string">		a sorted list of tuples (term, p_val)</span></span><br><span class="line"><span class="string">	"""</span></span><br><span class="line">	<span class="comment">## check input:</span></span><br><span class="line">	<span class="keyword">if</span> <span class="keyword">not</span> gmt_fn.endswith(<span class="string">'.gmt'</span>):</span><br><span class="line">		<span class="keyword">raise</span> IOError(<span class="string">"The gene-set library file is not in GMT format"</span>)</span><br><span class="line"></span><br><span class="line">	<span class="comment">## read gmt into a dict:</span></span><br><span class="line">	res = []</span><br><span class="line">	<span class="keyword">with</span> open (gmt_fn) <span class="keyword">as</span> f:</span><br><span class="line">		<span class="keyword">for</span> line <span class="keyword">in</span> f:</span><br><span class="line">			sl = line.strip().split(<span class="string">'\t'</span>)</span><br><span class="line">			term = sl[<span class="number">0</span>]</span><br><span class="line">			genes = sl[<span class="number">2</span>:]</span><br><span class="line">			<span class="keyword">if</span> <span class="string">','</span> <span class="keyword">in</span> genes[<span class="number">0</span>]: <span class="comment">## auto detect fuzzy gmts</span></span><br><span class="line">				genes = [gene.split(<span class="string">','</span>)[<span class="number">0</span>] <span class="keyword">for</span> gene <span class="keyword">in</span> genes]</span><br><span class="line">			principal_angle, p_val = paea(chdir, genes, case_sensitive=case_sensitive)</span><br><span class="line">			res.append( (term, p_val) )</span><br><span class="line">	<span class="keyword">if</span> sort:<span class="comment">## sort terms based on p values in ascending order</span></span><br><span class="line">		res = sorted(res, key=<span class="keyword">lambda</span> x:x[<span class="number">1</span>])</span><br><span class="line">	<span class="keyword">else</span>: <span class="comment">## if not sort, return unsorted p_vals only</span></span><br><span class="line">		res = [p_val <span class="keyword">for</span> term, p_val <span class="keyword">in</span> res ]</span><br><span class="line">	<span class="keyword">return</span> res</span><br><span class="line"></span><br><span class="line"><span class="keyword">from</span> collections <span class="keyword">import</span> OrderedDict</span><br><span class="line"><span class="comment">### Create dataframe and dictionary to store the results</span></span><br><span class="line">cd_results = pd.DataFrame(index=BRCA_RNA_df_no_outliers.index)</span><br><span class="line">cd_genes = &#123;&#125;</span><br><span class="line"></span><br><span class="line"><span class="comment"># Conduct the DGE Analysis</span></span><br><span class="line"><span class="comment"># Based on different PAM50 Groups</span></span><br><span class="line"></span><br><span class="line"><span class="comment"># Get the PAM50 for each sample in a dictionary</span></span><br><span class="line">sample_PAM50 = OrderedDict(principalDf_[<span class="string">'PAM50 mRNA'</span>].to_dict())</span><br><span class="line">comparison = <span class="string">'Luminal_A_vs_B'</span></span><br><span class="line">mapping_dict = &#123;<span class="string">'Basal-like'</span>: <span class="number">0</span>, <span class="string">'HER2-enriched'</span>: <span class="number">0</span>, <span class="string">'Luminal A'</span>: <span class="number">1</span>, <span class="string">'Luminal B'</span>:<span class="number">2</span>&#125;</span><br><span class="line"></span><br><span class="line"><span class="comment"># 1. Luminal A vs B</span></span><br><span class="line"><span class="comment">### </span></span><br><span class="line">sampleclass = np.array([mapping_dict[P] <span class="keyword">for</span> P <span class="keyword">in</span> principalDf_[<span class="string">'PAM50 mRNA'</span>].tolist()])</span><br><span class="line"></span><br><span class="line"><span class="comment">### Conduct CHDIR analysis</span></span><br><span class="line">cd_res = chdir(BRCA_RNA_df_no_outliers.values, sampleclass, BRCA_RNA_df_no_outliers.index)</span><br><span class="line"></span><br><span class="line">print(cd_res)</span><br><span class="line"></span><br><span class="line">cd_df = pd.DataFrame(&#123;<span class="string">'Gene'</span>: [x[<span class="number">1</span>] <span class="keyword">for</span> x <span class="keyword">in</span> cd_res], <span class="string">'cd_value'</span>: [x[<span class="number">0</span>] <span class="keyword">for</span> x <span class="keyword">in</span> cd_res]&#125;)</span><br><span class="line">cd_df.index = cd_df[<span class="string">'Gene'</span>]</span><br><span class="line">cd_df.drop(<span class="string">'Gene'</span>, axis=<span class="number">1</span>, inplace=<span class="literal">True</span>)</span><br><span class="line">cd_df.head()</span><br><span class="line"></span><br><span class="line"><span class="comment"># sort CD in by absolute values in descending order</span></span><br><span class="line">up_genes = cd_df[<span class="string">'cd_value'</span>].abs().sort_values().index[:<span class="number">500</span>]</span><br><span class="line">dn_genes = cd_df[<span class="string">'cd_value'</span>].abs().sort_values().index[<span class="number">-500</span>:]</span><br><span class="line"></span><br><span class="line"><span class="comment"># split up and down</span></span><br><span class="line">cd_genes[comparison+<span class="string">'-up'</span>] = up_genes</span><br><span class="line">cd_genes[comparison+<span class="string">'-dn'</span>] = dn_genes</span><br></pre></td></tr></table></figure>

<!-- Tags -->



<div class="tags">
    
</div>



<!-- Comments -->
<div>
    


</div>



            </div>
        </div>

        <!-- Footer -->
<footer id="footer">
    <div class="inner">
        <section>
            <h2>About</h2>
            <div>
                Just for fun
            </div>
        </section>
        <section>
            <h2>Follow</h2>
            <ul class="icons">
                
                
                
                
                
                    <li><a href="https://github.com/klugjo/hexo-theme-phantom/commits?author=klugjo" class="icon style2 fa-github" target="_blank" ><span class="label">GitHub</span></a></li>
                
                
                
                
                
                    <li><a href="hebeeeeeee@outlook.com" class="icon style2 fa-envelope-o" target="_blank" ><span class="label">Email</span></a></li>
                
                
            </ul>
        </section>
        <ul class="copyright">
            <li>&copy; Untitled. All rights reserved</li>
            <li>Design: <a href="http://html5up.net" target="_blank">HTML5 UP</a></li>
            <li>Hexo: <a href="http://www.codeblocq.com/" target="_blank">Jonathan Klughertz</a></li>
        </ul>
    </div>
</footer>
    </div>

    <!-- After footer scripts -->
    
<!-- jQuery -->
<script src="/js/jquery.min.js"></script>

<!-- skel -->
<script src="/js/skel.min.js"></script>

<!-- Custom Code -->
<script src="/js/util.js"></script>

<!--[if lte IE 8]>
<script src="/js/ie/respond.min.js"></script>
<![endif]-->

<!-- Custom Code -->
<script src="/js/main.js"></script>

<!-- Gallery -->
<script src="//cdn.rawgit.com/noelboss/featherlight/1.3.5/release/featherlight.min.js" type="text/javascript" charset="utf-8"></script>

<!-- Disqus Comments -->


</body>

</html>